mParticle/mparticle-web-sdk
# feat: Add Offline Storage for Events and Batches via Feature Flag

## Instructions
 1. PR target branch should be against `development`
 2. PR title name should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-title-check.yml
 3. PR branch prefix should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-branch-check-name.yml

 ## Summary
* Introduces a refactor of our Web API Storage Wrapper (Vault) to support SessionStorage and LocalStorage
* Implements two instances of Vaults; one for Events (Session Storage), and a second for Batches (Local Storage)
* As events are queued, they are placed into Session Storage for later processing. Once the `prepareAndUpload` method is triggered, events are removed from Session Storage and converted to Batches for upload.
* If any Batches exist in Local Storage, they are prepended to the in-memory Batch Queue, then purged from Local Storage to prevent accidental retransmissions.
  * Note: In cases where a Beacon Upload is requested, Batches are stored in Local Storage before transmission, as it is a "fire-and-forget" upload.
* If the transmission is successful, nothing is stored in Local Storage and the in-memory Batch Queue is purged.
* If the transmission is unsuccessful, the batches are saved to Local Storage and are retransmitted during the next interval, in the order they were created.
* This new logic is wrapped in a Feature Flag so we can control the release of Offline Storage via a config flag.
* Also includes a small refactor of MockBatchCreator for testing and allows console logs to appear in `npm run test:debug`

 ## Testing Plan
 - Run automated tests
 - Verify manually using a sample app
   - Events should be stored in Session Storage before conversion to batches
   - Events should be cleared from Session Storage once converted to batches
   - Batches should not be stored in Local Storage before upload, unless `useBeacon` is true
   - Restarting the browser before events are converted to batches should retain events within Session Storage
   - Batches that do not upload should be put back into local storage, in the order they were created
   - Reloading the sample app should reload batches and events from Local Storage and Session Storage, respectively
   - Events and Batches should maintain sequence at all points during the batch upload process
   - Inspect Live Stream for any potential duplicate events or batches

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://go.mparticle.com/work/SQDSDKS-5171
